### PR TITLE
WIP: Add shortcut for normal favorite location

### PR DESCRIPTION
**Pull request by Noxsense**
Hello, I'm currently working on #329.
I tried to put the your new `addShortcut()` into the parent class, so the `FavoriteTripPopupMenu` can also create shortcuts, if they are marked as favorites.
I used the logic of `findDirection(...)` for the Shortcut Intent and checked if all parameters were correctly inserted, but the shortcut is only created (or added) if the parameters for `from`, `via` and `to` are null (uid and boolean were fine).

(The Intent created for the shortcut is actually working: I tried to start its activity instead of putting it as shortcut into the other intent.)

I have no idea the shortcut isn't showing up or only if the  parameters were null.
Do you have any idea why this happens?

### CODE REVIEW

File: app/src/main/java/de/grobox/transportr/utils/TransportrUtils.java [hunk 1]
```diff
@@ -649,6 +649,22 @@ static public int getDragDistance(Context context) {
...
+		shortcutIntent.setAction(ACTION_MAIN);
+		shortcutIntent.addFlags(FLAG_ACTIVITY_NEW_TASK);
+		shortcutIntent.addFlags(FLAG_ACTIVITY_CLEAR_TOP);
+
+		//FIXME Doesn't work if they are not null
+		shortcutIntent.putExtra(FAV_TRIP_UID, uid);
+		shortcutIntent.putExtra(FROM, from);
+		shortcutIntent.putExtra(VIA, via);
+		shortcutIntent.putExtra(TO, to);
+		shortcutIntent.putExtra(SEARCH, false);
```
Comment (grote) [inline app/src/main/java/de/grobox/transportr/utils/TransportrUtils.java]: My guess is that you can't put complex objects that need to be serialized into shortcut intents.

Comment (grote) [reply app/src/main/java/de/grobox/transportr/utils/TransportrUtils.java]: If that's true, you could work with `StoredSearch` `uid`s, but you need to handle the case when the `uid` doesn't exist anymore.

**Comment by Noxsense** [comment]
Hey, I've tried to store the locations ID ((instead of only StoredSearches)) and I could restore those from the intent.
But I think it' because the shortcut is trying to start searching immediately the database is not always loaded at time an there are no locations to restore from yet.
I used `getLocations` from the `DirectionViewModel` in the `DirectionActivity` but also the database itself isn't always loaded yet.

It's working if the shortcut is launched not long after quitting the app, I guess it's because of the cache.

Then I tried to load the already existing Home shortcut with cleared cache (or just updated app) and there the locations doesn't load either.

File: app/src/main/java/de/grobox/transportr/favorites/trips/FavoriteTripPopupMenu.java [hunk 2]
```diff
@@ -50,12 +50,14 @@ public boolean onMenuItemClick(MenuItem item) {
...
-				widgetName = String.format(widgetName, String.valueOf(trip.getFrom()), String.valueOf(trip.getTo()));
-
-				addShortcut(context, getShortcutIntent(context, 0, trip.getFrom(), trip.getVia(), trip.getTo()),
-						widgetName, R.mipmap.ic_quickwork_widget);
+				addShortcut(context, getShortcutIntent(context,
+							trip.getUid(),
+							trip.getNetworkId(),
+							((StoredLocation) trip.getFrom()).getUid(),
+							(trip.getVia() == null) ? null : ((StoredLocation) trip.getVia()).getUid(),
+							((StoredLocation) trip.getTo()).getUid()),
```
Comment (grote) [inline app/src/main/java/de/grobox/transportr/favorites/trips/FavoriteTripPopupMenu.java]: `trip.getUid()` should be sufficient to identify a trip and get the `FavoriteTripItem` from the database. No need for the other data here.

File: app/src/main/java/de/grobox/transportr/favorites/trips/FavoriteTripPopupMenu.java [hunk 3]
```diff
@@ -50,12 +50,14 @@ public boolean onMenuItemClick(MenuItem item) {
...
+						getShortcutName(),
+						R.mipmap.ic_quickwork_widget);
```
Comment (grote) [inline app/src/main/java/de/grobox/transportr/favorites/trips/FavoriteTripPopupMenu.java]: When the PR is ready, we would need to think about  a different icon here.

File: app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java [hunk 4]
```diff
@@ -126,6 +128,30 @@ private void processIntent() {
...
+					fromId = intent.getLongExtra(FROM, 0);
+					viaId = (Long) intent.getSerializableExtra(VIA);
+					toId = intent.getLongExtra(TO, 0);
+
+					from = null;
+					to = null;
+
+					try{
+						for (FavoriteLocation location : viewModel.getLocations().getValue()) {
+							locationId = location.getUid();
```
Comment (grote) [inline app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java]: You need to ask the view model for the FavoriteTripItem with the `uid` from the intent. If the data is not yet available when the intent is processed, you need to observe it to get notified when available.

File: app/src/main/java/de/grobox/transportr/utils/TransportrUtils.java [hunk 5]
```diff
@@ -649,19 +650,18 @@ static public int getDragDistance(Context context) {
...
 		shortcutIntent.putExtra(FAV_TRIP_UID, uid);
-		shortcutIntent.putExtra(FROM, from);
-		shortcutIntent.putExtra(VIA, via);
-		shortcutIntent.putExtra(TO, to);
-		shortcutIntent.putExtra(SEARCH, false);
-
+		shortcutIntent.putExtra("network", networkId.ordinal());
+		shortcutIntent.putExtra(FROM, fromId);
+		shortcutIntent.putExtra(VIA, viaId);
+		shortcutIntent.putExtra(TO, toId);
```
Comment (grote) [inline app/src/main/java/de/grobox/transportr/utils/TransportrUtils.java]: like I said above, the `FAV_TRIP_UID` should be sufficient here.

**Comment by Noxsense** [comment]
The name for the shortcut is currently hard coded. But if all names from to and via have to be inserted, it will be quite long.
And I think the icon could the star like the favourite icon  on a red background.

**Comment by grote** [comment]
> The name for the shortcut is currently hard coded. But if all names from to and via have to be inserted, it will be quite long.

Yeah, I don't know how to solve that.

> And I think the icon could the star like the favourite icon on a red background.

I agree!

**Comment by comradekingu** [comment]
@Noxsense Could you post a screenshot of what that looks like?

**Comment by Noxsense** [comment]
I guess something like this?
ic_favourite_generated


[Icon (with Launcher Icon generator)](https://romannurik.github.io/AndroidAssetStudio/icons-launcher.html#foreground.type=clipart&foreground.clipart=star&foreground.space.trim=1&foreground.space.pad=0.15&foreColor=rgb(255%2C%20255%2C%20255)&backColor=rgb(242%2C%2066%2C%2054)&crop=0&backgroundShape=square&effects=shadow&name=ic_favorite_widget)

**Comment by grote** [comment]
Yes looks nice. Android Studio has a built-in icon creator now btw.

**Comment by Noxsense** [comment]
Is there anything else I can improve?

File: app/src/main/java/de/grobox/transportr/favorites/trips/AbstractFavoritesPopupMenu.java [hunk 6]
```diff
@@ -64,4 +71,22 @@ public boolean onMenuItemClick(MenuItem item) {
 		}
 	}
 
+	protected void addShortcut(Context context, Intent shortcutIntent, String shortcutName, int shortcutDrawable) {
```
Comment (grote) [inline app/src/main/java/de/grobox/transportr/favorites/trips/AbstractFavoritesPopupMenu.java]: Do you really need to pass in a `Context` here or can you use the existing `context` variable?

The `shortcutDrawable` should be annotated with `@DrawableRes`.

File: app/src/main/java/de/grobox/transportr/favorites/trips/AbstractFavoritesPopupMenu.java [hunk 7]
```diff
@@ -64,4 +71,22 @@ public boolean onMenuItemClick(MenuItem item) {
...
+		context.sendBroadcast(addIntent);
+
+		// switch to home-screen to let the user see the new shortcut
+		Intent startMain = new Intent(Intent.ACTION_MAIN);
+		startMain.addCategory(Intent.CATEGORY_HOME);
+		startMain.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+		context.startActivity(startMain);
+	}
+
+	protected abstract Intent getShortcutIntent(Context context, String type);
```
Comment (grote) [inline app/src/main/java/de/grobox/transportr/favorites/trips/AbstractFavoritesPopupMenu.java]: Why is this needed if you pass in the shortcut intent above?

File: app/src/main/java/de/grobox/transportr/favorites/trips/FavoriteTripPopupMenu.java [hunk 8]
```diff
@@ -20,17 +20,27 @@
...
 import android.view.View;
 
 import de.grobox.transportr.R;
+import de.grobox.transportr.trips.search.DirectionsActivity;
+
+import static android.content.Intent.ACTION_SEARCH;
+import static android.content.Intent.FLAG_ACTIVITY_NEW_TASK;
+import static android.content.Intent.FLAG_ACTIVITY_CLEAR_TOP;
+
+//import static de.grobox.transportr.utils.Constants.FAV_TRIP_UID;
```
Comment (grote) [inline app/src/main/java/de/grobox/transportr/favorites/trips/FavoriteTripPopupMenu.java]: Is this still needed?

File: app/src/main/java/de/grobox/transportr/favorites/trips/FavoriteTripPopupMenu.java [hunk 9]
```diff
@@ -67,4 +80,32 @@ private void setFavState(MenuItem item, boolean is_fav) {
...
+	@Override
+	protected Intent getShortcutIntent(Context context, String type)
+	{
+		// ignore type, insert uid
+		Intent shortcutIntent = new Intent(context, DirectionsActivity.class);
+		shortcutIntent.setAction(ACTION_SEARCH);
+		shortcutIntent.addFlags(FLAG_ACTIVITY_NEW_TASK);
+		shortcutIntent.addFlags(FLAG_ACTIVITY_CLEAR_TOP);
+		shortcutIntent.setData(Uri.parse(DirectionsActivity.INTENT_URI_FAVORITE));
+		shortcutIntent.putExtra("uid", trip.getUid());
```
Comment (grote) [inline app/src/main/java/de/grobox/transportr/favorites/trips/FavoriteTripPopupMenu.java]: `"uid"` should be a `String` constant.

File: app/src/main/java/de/grobox/transportr/favorites/trips/FavoriteTripPopupMenu.java [hunk 10]
```diff
@@ -67,4 +80,32 @@ private void setFavState(MenuItem item, boolean is_fav) {
...
+		return shortcutIntent;
+	}
+
+	private String getShortcutName()
+	{
```
Comment (grote) [inline app/src/main/java/de/grobox/transportr/favorites/trips/FavoriteTripPopupMenu.java]: coding style violation (here and below in this method)

File: app/src/main/java/de/grobox/transportr/favorites/trips/FavoriteTripPopupMenu.java [hunk 11]
```diff
@@ -67,4 +80,32 @@ private void setFavState(MenuItem item, boolean is_fav) {
...
+		String str = null;
+
+		if (trip.getVia()==null)
```
Comment (grote) [inline app/src/main/java/de/grobox/transportr/favorites/trips/FavoriteTripPopupMenu.java]: missing spaces

File: app/src/main/java/de/grobox/transportr/favorites/trips/SavedSearchesViewModel.java [hunk 12]
```diff
@@ -42,7 +42,7 @@ protected SavedSearchesViewModel(TransportrApplication application, TransportNet
 		this.savedSearches = searchesRepository.getFavoriteTrips();
 	}
 
-	LiveData<List<FavoriteTripItem>> getFavoriteTrips() {
+	public LiveData<List<FavoriteTripItem>> getFavoriteTrips() {
```
Comment (grote) [inline app/src/main/java/de/grobox/transportr/favorites/trips/SavedSearchesViewModel.java]: Why did you need to change the visibility here? Do you access this ViewModel outside of this package?

File: app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java [hunk 13]
```diff
@@ -147,6 +152,26 @@ private void processIntent() {
 					to = viewModel.getWork().getValue();
 					if (to == null) new WorkPickerFragment().show(getSupportFragmentManager(), WorkPickerFragment.TAG);
 					break;
+				case INTENT_URI_FAVORITE:
+					from = null;
+					via = null;
+					to = null;
```
Comment (grote) [inline app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java]: Do you need to explicitly assign `null` here?

File: app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java [hunk 14]
```diff
@@ -147,6 +152,26 @@ private void processIntent() {
...
+					long uid = intent.getLongExtra("uid",0);
```
Comment (grote) [inline app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java]: Missing space. Btw. in Android Studio, there's a shortcut to automatically apply the current coding style to the entire file or your selection. This quickly fixes the style.

File: app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java [hunk 15]
```diff
@@ -147,6 +152,26 @@ private void processIntent() {
...
+
+					LiveData<List<FavoriteTripItem>> tripsLiveData;
+					tripsLiveData = viewModel.getFavoriteTrips();
+					tripsLiveData.observe(this, trips -> {
+						for (FavoriteTripItem trip : trips) {
+							if (trip!=null && trip.getUid()==uid)
+							{
```
Comment (grote) [inline app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java]: coding style

File: app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java [hunk 16]
```diff
@@ -147,6 +152,26 @@ private void processIntent() {
...
 					break;
+				case INTENT_URI_FAVORITE:
+					from = null;
+					via = null;
+					to = null;
+					long uid = intent.getLongExtra("uid",0);
+
+					LiveData<List<FavoriteTripItem>> tripsLiveData;
+					tripsLiveData = viewModel.getFavoriteTrips();
+					tripsLiveData.observe(this, trips -> {
```
Comment (grote) [inline app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java]: You could just do `viewModel.getFavoriteTrips().observe(...)`

File: app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java [hunk 17]
```diff
@@ -147,6 +152,26 @@ private void processIntent() {
...
+							if (trip!=null && trip.getUid()==uid)
+							{
+								searchFromTo(trip.getFrom(), trip.getVia(), trip.getTo());
+								tripsLiveData.removeObservers(this); // everything is done
+								break;
+							}
+						}
+					});
+					setIntent(null);
+					return; // will be finished in viewmodel's loaded trips
```
Comment (grote) [inline app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java]: This looks a bit strange. Why don't you let this method go down to the end and assign `from`, `via` and `to` properly instead of suddenly returning here and duplicating code before?

File: app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java [hunk 18]
```diff
@@ -147,6 +152,26 @@ private void processIntent() {
...
+					tripsLiveData = viewModel.getFavoriteTrips();
+					tripsLiveData.observe(this, trips -> {
+						for (FavoriteTripItem trip : trips) {
+							if (trip!=null && trip.getUid()==uid)
+							{
+								searchFromTo(trip.getFrom(), trip.getVia(), trip.getTo());
+								tripsLiveData.removeObservers(this); // everything is done
+								break;
+							}
+						}
```
Comment (grote) [inline app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java]: What happens if the user deleted a favorite trip, but still has the shortcut on the launcher? How is this handled?

File: app/src/main/java/de/grobox/transportr/favorites/trips/SavedSearchesViewModel.java (hunk 12 above)
Comment (Noxsense) [reply app/src/main/java/de/grobox/transportr/favorites/trips/SavedSearchesViewModel.java]: Yes, in `DirectionsActivity`
where I use it like `viewModel.getFavoriteTrips().observe(...)`

File: app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java (hunk 16 above)
Comment (Noxsense) [reply app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java]: Yes, but I use `tripsLiveData` to remove the observer when the actual trip is found.
Otherwise the found trip will be found again on every update of the livedata with the same trips.

File: app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java (hunk 18 above)
Comment (Noxsense) [reply app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java]: The shortcut will just start a `DirectionsActivity` with empty source and destination fields. The observer won't find a matching trip in the list and do nothing.
Should we try to delete shortcuts when unmarking trips as favorite?

File: app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java (hunk 17 above)
Comment (mheinzel) [reply app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java]: The main problem is that we don't know when all trips will be loaded and if they even contain the one we are looking for. So if we just use `getValue()` (as for Home or Work) it might not be loaded in time and we miss it. But we can't block here and wait for it either, since it might never appear.
The only solution I saw, was to attach `searchFromTo` to the observer and exit the method. But I don't have a valid `from` and `to` at this point, so I can't let the method go down to the end.

If you have a better idea, I would be happy to hear it.

I'll fix the other issues you pointed out.

Comment (mheinzel) [reply app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java]: Oh, and thank you for taking the time to review this!

File: app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java (hunk 16 above)
Comment (grote) [reply app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java]: Ah right. You need a reference to the LiveData.

File: app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java (hunk 18 above)
Comment (grote) [reply app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java]: I think when the observer doesn't find any trips, we could just show a Toast saying that this favorite was removed and that the user should remove the shortcut as well.

File: app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java [hunk 19]
```diff
@@ -147,6 +152,26 @@ private void processIntent() {
...
+					from = null;
+					via = null;
+					to = null;
+					long uid = intent.getLongExtra("uid",0);
+
+					LiveData<List<FavoriteTripItem>> tripsLiveData;
+					tripsLiveData = viewModel.getFavoriteTrips();
+					tripsLiveData.observe(this, trips -> {
+						for (FavoriteTripItem trip : trips) {
+							if (trip!=null && trip.getUid()==uid)
```
Comment (grote) [inline app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java]: `trip` should never be `null` here.

File: app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java (hunk 17 above)
Comment (Noxsense) [reply app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java]: The main problem is that we don't know when all trips will be loaded and
if they even contain the one we are looking for. So if we just use
`getValue()` (as for Home or Work) it might not be loaded in time and we
miss it. But we can't block here and wait for it either, since it might
never appear.
The only solution I saw, was to attach `searchFromTo` to the observer
and exit the method. But I don't have a valid `from` and `to` at this
point, so I can't let the method go down to the end.

If you have a better idea, I would be happy to hear it.

Thank you for taking the time to review my changes.
I'll fix the other
issues you pointed out.

Comment (grote) [reply app/src/main/java/de/grobox/transportr/trips/search/DirectionsActivity.java]: Ah I see. Unfortunately, I don't see a better way to handle this myself.

File: app/src/main/java/de/grobox/transportr/favorites/trips/AbstractFavoritesPopupMenu.java (hunk 7 above)
Comment (Noxsense) [reply app/src/main/java/de/grobox/transportr/favorites/trips/AbstractFavoritesPopupMenu.java]: Previously, it was necessary to call `getShortcutIntent()` to get the intent and pass it to `addShortcut()`.
This is unnecessary, so I changed `addShortcut()` to call `getShortcut()` itself.

`getShortcutIntent()` is needed because each subclass has its own way of creating the actual shortcut intent.

File: app/src/main/java/de/grobox/transportr/favorites/trips/FavoriteTripPopupMenu.java [hunk 20]
```diff
@@ -47,6 +57,9 @@ public boolean onMenuItemClick(MenuItem item) {
 				}
 				setFavState(item, trip.isFavorite());
 				return true;
+			case R.id.action_add_shortcut:
+				addShortcut(null, getShortcutName(), R.mipmap.ic_launcher_favorite);
```
Comment (grote) [inline app/src/main/java/de/grobox/transportr/favorites/trips/FavoriteTripPopupMenu.java]: What is a `null` type? If `null` makes sense here, you need to annotate the parameter `@Nullable`.

File: app/src/main/java/de/grobox/transportr/favorites/trips/FavoriteTripPopupMenu.java [hunk 21]
```diff
@@ -67,4 +80,21 @@ private void setFavState(MenuItem item, boolean is_fav) {
...
+	protected Intent getShortcutIntent(Context context, String type) {
+		// ignore type, insert uid
+		Intent shortcutIntent = new Intent(context, DirectionsActivity.class);
+		shortcutIntent.setAction(ACTION_SEARCH);
+		shortcutIntent.addFlags(FLAG_ACTIVITY_NEW_TASK);
+		shortcutIntent.addFlags(FLAG_ACTIVITY_CLEAR_TOP);
+		shortcutIntent.setData(Uri.parse(DirectionsActivity.INTENT_URI_FAVORITE));
+		shortcutIntent.putExtra(FAV_TRIP_UID, trip.getUid());
+		return shortcutIntent;
+	}
```
Comment (grote) [inline app/src/main/java/de/grobox/transportr/favorites/trips/FavoriteTripPopupMenu.java]: Could you also pass in the proper type here and then do

```java
Intent shortcutIntent = super.getShortcutIntent(Context context, String type);
shortcutIntent.putExtra(FAV_TRIP_UID, trip.getUid());
return shortcutIntent;
```

This way, you can maybe get rid of the code duplication.

Comment (Noxsense) [reply app/src/main/java/de/grobox/transportr/favorites/trips/FavoriteTripPopupMenu.java]: Ok, now I am reusing some more of the methods which where actually in `SpecialLocationPopupMenu` in `AbstractFavoritesPopupMenu`, while most of them are still abstract and are overridden in the actual sub class. So all of them use `addShortcut()` and `getShortcutIntent()` with `FavoriteTripPopupMenu` adding the trips `uid`.

**Comment by grote** [comment]
Merged, thanks!

**Comment by Noxsense** [comment]
Nice. Thanks for your help!
